### PR TITLE
build documentation in doxygen

### PR DESCRIPTION
**Pull request by balazsracz**
Adds documentation about the OpenMRN build system.

### CODE REVIEW

File: doc/build_system.dox [hunk 1]
```diff
@@ -0,0 +1,549 @@
...
+there.
+
+The most important concept is the *target*.
+- It defines the toolchain and compiler settings to be used;
+- defines which operating system the code shall be compiled for;
+- sets all the compilation flags;
+- defines the code that shall be compiled vs skipped.
+
+Generally, all code in `openmrn/src` gets compiled once per *core target*, and
+all code in the application tree gets compiled once per *application target*.
```
Comment (bakerstu) [inline doc/build_system.dox]: you can make *core target* and *application target* @ref hyperlinks, e.g. @ref core_target

File: doc/build_system.dox [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+  - {@code ti-cc3200-launchpad} Pin definitions and hardware initialization
+    code for the CC3200-LAUNCHXL launchpad. References files from the
+    `ti-cc3200-generic` directory.
+
+- {@code doc} The documentation using Doxygen, and the make scripts to
+  generate it. Run `make docs` in this directory to generate the HTML output of
+  the documentation.
+
+- {@code etc} Makefile libraries that are used by the build system. The `*.mk`
+  files here get included into the makefiles.
```
Comment (bakerstu) [inline doc/build_system.dox]: Capital "M" Makefiles?

File: doc/build_system.dox [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+    FreeRTOS, and to some extent windows. The concepts include threads,
+    mutexes, sempahores, timers, support for select, GPIOs, mDNS.
+
+  - {@code utils} A large number of diverse utility classes for data
+    structures, buffers & queues, logging, testing, printing and formatting
+    data, creating and using streams and sockets. Helper classes for
+    implementing data hubs / repeaters, including tying them to various device
+    drivers such as CAN-bus or USB.
+
+  - {@code withrottle} Implementation for JMRI's wireless throttle protocol.\
```
Comment (bakerstu) [inline doc/build_system.dox]: Maybe a note that WiThrottle support is incomplete

File: doc/build_system.dox [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+compilation is a single `.a` archive, for example `openlcb.a`. There are a
+couple of deprecated (empty) core target directories which need to be deleted,
+too, such as `cue`, `if`, `core`.
+
+- {@code lib} Contains all the core library archive files. This is the place
+  where the application targets will go look for the core libraries for
+  linking.
+
+- {@code cJSON, tinyxml2} These are dependent libraries that are redistributed
+  with the OpenMRN code. They get compiled like a core library.
```
Comment (bakerstu) [inline doc/build_system.dox]: We should probably remove these if they are not being used.

File: doc/build_system.dox [hunk 5]
```diff
@@ -0,0 +1,549 @@
...
+
+- https://github.com/bakerstu/openmrn/tree/master/applications/io_board
+- https://github.com/bakerstu/openmrn/tree/master/applications/hub
+- https://github.com/balazsracz/openmrn_cue/tree/master/cs , which is an example
+of an externally shipped application.
+
+@subsubsection app_dir_location Locating application and OpenMRN directory
+
+An application directory can be located anywhere in the file system. The
+compilation can be started inside the application's toplevel directory (will
```
Comment (JohnSL) [inline doc/build_system.dox]: Change "toplevel" to "top-level"

File: doc/build_system.dox [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+appropriate absolute directory path. Everything inside the OpenMRN tree will
+then be referenced from this path, including makefile helper libraries, finding
+the core targets and their compilation outputs, include path for `*.h` and
+`*.hxx` files, etc... Usually at the beginning of the Makefile we set
+OPENMRNPATH if it's not already set in the environment. For applications that
+are shipped within the OpenMRN tree we just set `OPENMRNPATH=$(realpath ..)`
+(it's important that this path be absolute and not relative since we'll refer
+to it from different places), applications that are placed elsewhere in the
+filesystem need to set OPENMRNPATH explicitly or in the user startup script
+(e.g., in `~/.bashrc` writing `export OPENMRNPATH=~/openmrn`).
```
Comment (JohnSL) [inline doc/build_system.dox]: Should probably mention that you can't use ~ in the Makefile itself.

File: doc/build_system.dox [hunk 7]
```diff
@@ -0,0 +1,549 @@
...
+
+From the toplevel application directory we have the following files and
+subdirectories:
+
+- {@code Makefile} This is usually always the same, just copy an example.
+
+- {@code subdirs} Describes the list of application libraries.
+
+- {@code config.mk} By convention, included at the top of all target Makefiles
+  to defince common build settings for all targets. This is typically where we
```
Comment (JohnSL) [inline doc/build_system.dox]: Change "defince" to "define", unless defince is a term I don't know yet :-).

File: doc/build_system.dox [hunk 8]
```diff
@@ -0,0 +1,549 @@
...
+
+- {@code automata
+commandstation
+mobilestation
+custom
+server
+src}
+These are application library directories in the
+https://github.com/balazsracz/openmrn_cue/tree/master/cs application. They
+contain sources and headers for various aspects. All pplication libraries get
```
Comment (JohnSL) [inline doc/build_system.dox]: Change "pplication" to "application"

File: doc/build_system.dox [hunk 9]
```diff
@@ -0,0 +1,549 @@
...
+compiled once per application target, and supplied to the commandline as
+archives. If no object is instantiated in `main` from a specific application
+library, the linker will optimize and remove all of its code. Note that `src` is
+actually a library here as well, will compile into libsrc.a.
```
Comment (JohnSL) [inline doc/build_system.dox]: Change ", will" to ", and will"

Comment (balazsracz) [inline doc/build_system.dox]: done

File: doc/build_system.dox (hunk 8 above)
Comment (balazsracz) [inline doc/build_system.dox]: done

File: doc/build_system.dox (hunk 7 above)
Comment (balazsracz) [inline doc/build_system.dox]: done

File: doc/build_system.dox (hunk 6 above)
Comment (balazsracz) [inline doc/build_system.dox]: done

File: doc/build_system.dox (hunk 5 above)
Comment (balazsracz) [inline doc/build_system.dox]: done

File: doc/build_system.dox (hunk 4 above)
Comment (balazsracz) [inline doc/build_system.dox]: filed an issue to not forget about it

File: doc/build_system.dox [hunk 10]
```diff
@@ -0,0 +1,549 @@
...
+  - {@code withrottle} Implementation for JMRI's wireless throttle protocol.\
+
+- {@code targets} All core targets. These are listed in importance order, not
+alphabetically:
+
+  - {@code cov} Target for compiling and running unit tests (on the host) and
+    test coverage generation.
+
+  - {@code linux.x86} Host target using 'g++' as the compiler. This can be used
+    to compile for the host OS under linux. It also work on a Raspberry Pi.
```
Comment (balazsracz) [inline doc/build_system.dox]: Done, thanks. Next time if you want to write a shorter comment, just say
work->works

File: doc/build_system.dox (hunk 3 above)
Comment (balazsracz) [inline doc/build_system.dox]: Done (both).

File: doc/build_system.dox (hunk 2 above)
Comment (balazsracz) [inline doc/build_system.dox]: done

File: doc/build_system.dox [hunk 11]
```diff
@@ -0,0 +1,549 @@
...
+@section directory_structure Directory structure
+
+In the OpenMRN distribution we have the following structure:
+
+- {@code applications} Ccontains a number of example applications. The
+  application tree is described below.
+
+- {@code bin} Scripts and tools not directly used by the build system.
+
+- {@code boards} Hhardware-specific code that can be shared by different
```
Comment (balazsracz) [inline doc/build_system.dox]: Done. Also fixed Ccontains above.

File: doc/build_system.dox [hunk 12]
```diff
@@ -0,0 +1,549 @@
...
+The output of the compilation of a core target is a set of libraries and object
+files that can be linked into an application for that specific OS and
+processor.
+
+@subsection application_target Application target
+
+Application targets are found under `$ (APP_ROOT)/targets` directory, for
+example
+https://github.com/bakerstu/openmrn/tree/master/applications/hub/targets.  Each
+application target will turn into one binary. The above example shows that this
```
Comment (balazsracz) [inline doc/build_system.dox]: Documented next to the OPENMRNPATH discussion.

File: doc/build_system.dox (hunk 1 above)
Comment (balazsracz) [inline doc/build_system.dox]: done

File: doc/build_system.dox [hunk 13]
```diff
@@ -0,0 +1,549 @@
...
+
+@section core_concepts Concepts
+
+The OpenMRN build system is based upon recursive make. Each directory contains
+a Makefile, and during the compilation process these makefiles instruct the
+build program to recurse into the relevant subdirectories and execute `make`
+there.
+
+The most important concept is the *target*.
+- It defines the toolchain and compiler settings to be used;
```
Comment (balazsracz) [inline doc/build_system.dox]: I replaced them with commas.

**Pull request merged**